File: src/main/java/me/alegian/thavma/impl/init/data/providers/T7BlockTagProvider.java
package me.alegian.thavma.impl.init.data.providers;

import me.alegian.thavma.impl.Thavma;
import me.alegian.thavma.impl.common.block.InfusedBlock;
import me.alegian.thavma.impl.init.registries.T7Tags;
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks;
import net.minecraft.core.HolderLookup;
import net.minecraft.data.PackOutput;
import net.minecraft.tags.BlockTags;
import net.neoforged.neoforge.common.Tags;
import net.neoforged.neoforge.common.data.BlockTagsProvider;
import net.neoforged.neoforge.common.data.ExistingFileHelper;
import net.neoforged.neoforge.registries.DeferredHolder;
import org.jetbrains.annotations.Nullable;

import java.util.concurrent.CompletableFuture;

public class T7BlockTagProvider extends BlockTagsProvider {
  public T7BlockTagProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, @Nullable ExistingFileHelper existingFileHelper) {
    super(output, lookupProvider, Thavma.MODID, existingFileHelper);
  }

  @Override
  protected void addTags(HolderLookup.Provider pProvider) {
    var boundInfusedBlocksArray = T7Blocks.INSTANCE.getINFUSED_BLOCKS().stream().map(DeferredHolder::get).toArray(InfusedBlock[]::new);
    // TODO: add nitor
    this.tag(T7Tags.CrucibleHeatSourceTag.BLOCK).addTag(BlockTags.FIRE).addTag(BlockTags.CAMPFIRES);
    this.tag(Tags.Blocks.ORES).add(boundInfusedBlocksArray);
    this.tag(Tags.Blocks.ORE_RATES_SINGULAR).add(boundInfusedBlocksArray);
    this.tag(BlockTags.MINEABLE_WITH_PICKAXE).add(boundInfusedBlocksArray);
    this.tag(BlockTags.NEEDS_STONE_TOOL).add(boundInfusedBlocksArray);
    this.tag(BlockTags.NEEDS_IRON_TOOL).add(
        T7Blocks.INSTANCE.getARCANUM_BLOCK().get(),
        T7Blocks.INSTANCE.getORICHALCUM_BLOCK().get()
    );
    this.tag(BlockTags.MINEABLE_WITH_AXE).add(
        T7Blocks.INSTANCE.getGREATWOOD_LOG().get(),
        T7Blocks.INSTANCE.getSILVERWOOD_LOG().get(),
        T7Blocks.INSTANCE.getGREATWOOD_PLANKS().get(),
        T7Blocks.INSTANCE.getSILVERWOOD_PLANKS().get()
    );

    this.tag(BlockTags.LEAVES).add(T7Blocks.INSTANCE.getGREATWOOD_LEAVES().get(), T7Blocks.INSTANCE.getSILVERWOOD_LEAVES().get());
    this.tag(BlockTags.LOGS_THAT_BURN).add(T7Blocks.INSTANCE.getGREATWOOD_LOG().get(), T7Blocks.INSTANCE.getSILVERWOOD_LOG().get());
    this.tag(BlockTags.SAPLINGS).add(T7Blocks.INSTANCE.getGREATWOOD_SAPLING().get(), T7Blocks.INSTANCE.getSILVERWOOD_SAPLING().get());
    this.tag(BlockTags.PLANKS).add(T7Blocks.INSTANCE.getGREATWOOD_PLANKS().get(), T7Blocks.INSTANCE.getSILVERWOOD_PLANKS().get());

    this.tag(Tags.Blocks.STORAGE_BLOCKS).add(
        T7Blocks.INSTANCE.getARCANUM_BLOCK().get(),
        T7Blocks.INSTANCE.getORICHALCUM_BLOCK().get()
    );

    this.tag(BlockTags.BEACON_BASE_BLOCKS).add(
        T7Blocks.INSTANCE.getARCANUM_BLOCK().get(),
        T7Blocks.INSTANCE.getORICHALCUM_BLOCK().get()
    );
  }
}

File: src/main/java/me/alegian/thavma/impl/init/data/providers/T7BlockTagProvider.kt
package me.alegian.thavma.impl.init.data.providers

import me.alegian.thavma.impl.Thavma
import me.alegian.thavma.impl.init.registries.T7Tags.CrucibleHeatSourceTag.BLOCK
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.ARCANUM_BLOCK
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.GREATWOOD_LEAVES
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.GREATWOOD_LOG
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.GREATWOOD_PLANKS
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.GREATWOOD_SAPLING
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.INFUSED_DEEPSLATES
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.INFUSED_STONES
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.ORICHALCUM_BLOCK
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.SILVERWOOD_LEAVES
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.SILVERWOOD_LOG
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.SILVERWOOD_PLANKS
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.SILVERWOOD_SAPLING
import net.minecraft.core.HolderLookup
import net.minecraft.data.PackOutput
import net.minecraft.tags.BlockTags
import net.neoforged.neoforge.common.Tags
import net.neoforged.neoforge.common.data.BlockTagsProvider
import net.neoforged.neoforge.common.data.ExistingFileHelper
import java.util.concurrent.CompletableFuture

class T7BlockTagProvider(output: PackOutput, lookupProvider: CompletableFuture<HolderLookup.Provider?>, existingFileHelper: ExistingFileHelper?) : BlockTagsProvider(output, lookupProvider, Thavma.MODID, existingFileHelper) {
    override fun addTags(pProvider: HolderLookup.Provider) {
        for (infusedBlock in (INFUSED_STONES.values + INFUSED_DEEPSLATES.values)) {
            tag(Tags.Blocks.ORES).add(infusedBlock.get())
            tag(Tags.Blocks.ORE_RATES_SINGULAR).add(infusedBlock.get())
            tag(BlockTags.MINEABLE_WITH_PICKAXE).add(infusedBlock.get())
            tag(BlockTags.NEEDS_STONE_TOOL).add(infusedBlock.get())
        }

        // TODO: add nitor
        tag(BLOCK).addTag(BlockTags.FIRE).addTag(BlockTags.CAMPFIRES)
        tag(BlockTags.NEEDS_IRON_TOOL).add(
            ARCANUM_BLOCK.get(),
            ORICHALCUM_BLOCK.get()
        )
        tag(BlockTags.MINEABLE_WITH_AXE).add(
            GREATWOOD_LOG.get(),
            SILVERWOOD_LOG.get(),
            GREATWOOD_PLANKS.get(),
            SILVERWOOD_PLANKS.get()
        )

        tag(BlockTags.LEAVES).add(GREATWOOD_LEAVES.get(), SILVERWOOD_LEAVES.get())
        tag(BlockTags.LOGS_THAT_BURN).add(GREATWOOD_LOG.get(), SILVERWOOD_LOG.get())
        tag(BlockTags.SAPLINGS).add(GREATWOOD_SAPLING.get(), SILVERWOOD_SAPLING.get())
        tag(BlockTags.PLANKS).add(GREATWOOD_PLANKS.get(), SILVERWOOD_PLANKS.get())

        tag(Tags.Blocks.STORAGE_BLOCKS).add(
            ARCANUM_BLOCK.get(),
            ORICHALCUM_BLOCK.get()
        )

        tag(BlockTags.BEACON_BASE_BLOCKS).add(
            ARCANUM_BLOCK.get(),
            ORICHALCUM_BLOCK.get()
        )
    }
}
